driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/BatchCursorPublisher.java

```diff
@@ -116,44 +112,8 @@ public Integer getBatchSize() {
 
     @Override
     public void subscribe(final Subscriber<? super T> subscriber) {
-        batchCursor()
-                .flatMapMany(batchCursor -> {
```

Comment:
`flatMapMany` requests data immediately - causing the original issue.

Also there is no hook for setting the initial batchSize based on demand.